packages/snaps-execution-environments/src/offscreen/OffscreenSnapExecutor.ts

```diff
@@ -0,0 +1,166 @@
+   * request in the `data` property.
+   *
+   * @param data - The message data.
+   * @param data.data - The JSON-RPC request.
+   * @param data.jobId - The job ID.
+   * @param data.frameUrl - The URL to load in the iframe.
+   */
+  #onData(data: {
+    data: JsonRpcRequest<JsonRpcParams>;
+    jobId: string;
+    frameUrl: string;
+  }) {
+    const { jobId, frameUrl, data: request } = data;
+
+    if (!this.#jobs[jobId]) {
+      // This ensures that a job is initialized before it is used. To avoid
+      // code duplication, we call the `#onData` method again, which will
+      // run the rest of the logic after initialization.
+      this.#initializeJob(jobId, frameUrl)
+        .then(() => {
+          this.#onData(data);
+        })
+        .catch((error) => {
+          console.error('[Worker] Error initializing job:', error);
+        });
+
+      return;
+    }
+
+    // This is a method specific to the `OffscreenSnapExecutor`, as the service
+    // itself does not have access to the iframes directly.
+    if (request.method === 'terminateJob') {
+      this.#terminateJob(jobId);
+      return;
+    }
+
+    this.#jobs[jobId].stream.write(request);
+  }
+
+  /**
+   * Create a new iframe and set up a stream to communicate with it.
+   *
+   * @param jobId - The job ID.
+   * @param frameUrl - The URL to load in the iframe.
+   */
+  async #initializeJob(jobId: string, frameUrl: string): Promise<ExecutorJob> {
+    const window = await this.#createWindow(frameUrl, jobId);
+    const jobStream = new WindowPostMessageStream({
+      name: 'parent',
+      target: 'child',
+      targetWindow: window,
+      targetOrigin: '*',
+    });
+
+    // Write messages from the iframe to the parent, wrapped with the job ID.
+    jobStream.on('data', (data) => {
+      this.#stream.write({ data, jobId });
+    });
+
+    this.#jobs[jobId] = { id: jobId, window, stream: jobStream };
+    return this.#jobs[jobId];
+  }
+
+  /**
+   * Terminate the job with the given ID. This will close the iframe and delete
+   * the job from the internal job map.
+   *
+   * @param jobId - The job ID.
+   */
+  #terminateJob(jobId: string) {
+    assert(this.#jobs[jobId], `Job "${jobId}" not found.`);
+
+    const iframe = document.getElementById(jobId);
+    assert(iframe?.parentNode, `Iframe with ID "${jobId}" not found.`);
+
+    iframe.parentNode.removeChild(iframe);
+    this.#jobs[jobId].stream.destroy();
+    delete this.#jobs[jobId];
+  }
+
+  /**
+   * Creates the iframe to be used as the execution environment. This may run
+   * forever if the iframe never loads, but the promise should be wrapped in
+   * an initialization timeout in the SnapController.
+   *
+   * @param uri - The iframe URI.
+   * @param jobId - The job id.
+   * @returns A promise that resolves to the contentWindow of the iframe.
+   */
+  // TODO: Move this to a reusable utility, as it is also used in the iframe
```

Comment:
If we end up needing to do this - Can we move to a shared abstraction in this in this PR?